tools/functional-tester/etcd-tester/lease_stresser.go

```diff
@@ -0,0 +1,300 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"math/rand"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/etcdserver/api/v3rpc/rpctypes"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// time to live for lease
+const TTL = 15
+
+type leaseStressConfig struct {
+	numOfLeases  int
+	keysPerLease int
+}
+
+type leaseStresser struct {
+	endpoint string
+	cancel   func()
+	conn     *grpc.ClientConn
+	kvc      pb.KVClient
+	lc       pb.LeaseClient
+
+	success      int
+	failure      int
+	numOfLeases  int
+	keysPerLease int
+
+	// mu Mutex is used to protect the shared maps aliveLeases and revokedLeases
+	// which are accessed and modified by different go routines.
+	mu            sync.Mutex
+	aliveLeases   map[int64]struct{}
+	revokedLeases map[int64]struct{}
+
+	runWg   sync.WaitGroup
+	aliveWg sync.WaitGroup
+
+	hasSetup bool
+}
+
+type leaseStresserBuilder func(m *member) Stresser
+
+func newLeaseStresserBuilder(lsConfig *leaseStressConfig, lsTracker func(*leaseStresser)) leaseStresserBuilder {
+	return func(mem *member) Stresser {
+		ls := &leaseStresser{
+			endpoint:     mem.grpcAddr(),
+			numOfLeases:  lsConfig.numOfLeases,
+			keysPerLease: lsConfig.keysPerLease,
+		}
+		lsTracker(ls)
+		return ls
+	}
+}
+
+// setupOnce initialize the fields once beucase the leaseStresser is longliving and doesn't need to be re-initialize on each stress() call.
+func (ls *leaseStresser) setupOnce() error {
+	if ls.hasSetup {
+		return nil
+	}
+	if ls.numOfLeases == 0 {
+		panic("expect numOfLeases to be set")
+	}
+	if ls.keysPerLease == 0 {
+		panic("expect keysPerLease to be set")
+	}
+	conn, err := grpc.Dial(ls.endpoint, grpc.WithInsecure())
+	if err != nil {
+		return fmt.Errorf("%v (%s)", err, ls.endpoint)
+	}
+
+	ls.kvc = pb.NewKVClient(conn)
+	ls.lc = pb.NewLeaseClient(conn)
+	ls.conn = conn
+	ls.aliveLeases = make(map[int64]struct{})
+	ls.revokedLeases = make(map[int64]struct{})
+	ls.hasSetup = true
+	return nil
+}
+
+func (ls *leaseStresser) Stress() error {
+	plog.Infof("lease Stresser %v starting ...", ls.endpoint)
+	if err := ls.setupOnce(); err != nil {
+		return err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
+	ls.cancel = cancel
+	ls.runWg.Add(1)
+	go ls.run(ctx)
+	return nil
+}
+
+func (ls *leaseStresser) run(ctx context.Context) {
+	defer ls.runWg.Done()
+	ls.createLeases(ctx)
+	ls.randomlyDropLeases(ctx)
+}
```

Comment:
`ls.createLeases` will replace expired leases with new leases. Would that be sufficent for generating loads? we can bump of the number of leases created.